[console 17.0.0] record-header 自定义动作偶发丢 recordId:页头存在两种 DOM 变体,其中一种不注入请求体,端点回 missing_record_id

## 一句话

记录详情页头的自定义动作(record_header 挂载)点击后,平台负责往请求体注入当前记录 id(`recordId`)——但页头渲染存在**两种 DOM 变体**,其中一种变体下不注入,端点收到的请求体没有 recordId,回 `missing_record_id`。用户侧表现为「偶尔点一下没反应,再点一次就好」。

## 环境

- @objectstack/*@17.0.0(GA);下游业务项目(os-project-titanwind-ehr,私有)真机 Console 实测发现
- 与对象、与动作配置无关:该项目所有 record_header 自定义动作(多个对象、同一标准配方)均受影响

## 量化复现

同一颗页头【删除】动作按钮,同一记录页,连点 10 次:

- 9 次:请求体带 `recordId`,动作正常执行;
- 1 次:请求体**不带** `recordId` → 服务端 `missing_record_id`;
- 关键相关性:失败那次抓包时,**页头 DOM 层级与成功时不同**(两种变体与是否注入一一对应);
- 换一次页面渲染即恢复。

疑似与记录页头的 hydration/半加载窗口有关(页头在加载早期与稳定后渲染树不同,注入逻辑只挂在其中一种变体上)。

## 期望

两种页头渲染形态下,record_header 动作的 recordId 注入行为一致;或在拿不到记录 id 时禁用按钮而非发出残缺请求。

## 备注

下游项目按「平台能覆盖的问题不自造补丁」原则未做任何绕行;抓包与 DOM 对照证据存于该项目取证分支,需要可提供摘录。

<!-- os-dev-report -->

```json
{
  "issue": 5176,
  "status": "done",
  "branch": "claude/issue-5176-record-header-recordid",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5182",
  "premise_still_valid": true,
  "premise_note": "The DEFECT is real and reproduced deterministically. The card's MECHANISM framing is falsified: there is no second DOM variant carrying a second injection site. One dispatch path, one injection site; only the seed value varies with click timing. The reporter's 1:1 DOM correlation is a co-symptom (the same unresolved pageRecord withholds the title/highlights chrome AND starves the seed), not the cause.",
  "pm_assumption_check": {
    "several_distinct_paths_selected_by_timing": "FALSE — a record_header api action with recordIdParam always takes the same path (RecordDetailView apiHandler, absolute-target branch). The 4 sites the PM counted are not alternatives for this action: :709-727 is the legacy dataSource.update default branch, :786-788 the flow trigger, :835-846 the server-action bridge. Different action types, not a timing selection.",
    "failing_path_only_id_source_is_the_resolved_page_record": "TRUE, and this is the finding. The api path's chain was row -> literal recordId override -> pageRecord, and stopped. It never consulted pureRecordId (URL-derived, present on first render). The other three paths all DO fall back to pureRecordId — the api path was the odd one out.",
    "zone3_suggested_route_disable_the_affordance": "REJECTED on measurement. The record id is NOT unavailable in the failure window — pureRecordId is read off the URL synchronously; only the record BODY is pending. Disabling every header action for the duration of every record fetch would be a UX regression to work around an id the platform already holds. Took the ruling's FIRST limb (paths inject identically); the second limb is pinned where it genuinely applies (non-default recordIdField, which the URL cannot serve, now refuses instead of emitting an incomplete body)."
  },
  "root_cause": "isLoading (the only state the SkeletonDetail early return reads) is cleared in a route-keyed queueMicrotask, while pageRecord lands one dataSource.findOne round-trip later via a separate effect. For that whole window the real header renders with every record_header action live. Seeded from null, resolveRecordIdParamSeed abstains by design (no value AND no error — 'no row context' is documented as not that guard's business), so the caller injected nothing, raised nothing, and dispatched anyway. Emitted body was literally {} -> backend missing_record_id.",
  "summary": "Diagnosed the intermittent loss as a single-path timing defect, not two DOM variants, and fixed it by making the api dispatch path seed recordIdParam from pureRecordId (the URL record id) as a last resort — the same fallback the file's other three dispatch paths already use. Row stash and literal recordId override still outrank it; the retarget guard is preserved verbatim so a child-object action is never handed the parent page's id; an action keyed on a non-default recordIdField now takes the helper's named refusal rather than emitting an under-specified request. Adjacent residue reported and deliberately not patched: if pureRecordId were itself absent the chain would still end in silent abstention — unreachable on this route, so patching it would change behaviour for a case I could not reproduce.",
  "files_changed": [
    "packages/app-shell/src/views/RecordDetailView.tsx",
    "packages/app-shell/src/views/RecordDetailView.headerRecordIdLoadWindow.test.tsx",
    ".changeset/record-header-recordid-load-window-5176.md"
  ],
  "reproduction": "Deterministic, not a retry loop: dataSource.findOne returns a promise the test releases on demand, holding the failure window open; the harness captures the ActionProvider handlers built WHILE the record is in flight. The existing pin RecordDetailView.recordIdParamSeed.test.tsx waits for the capture whose context carries the LOADED record (context.record.id === RECORD_ID), which by construction sits after the window — that is why it never caught this.",
  "tests": "RED on unpatched main @072085deb: `pnpm exec vitest run packages/app-shell/src/views/RecordDetailView.headerRecordIdLoadWindow.test.tsx` -> 'Test Files 1 failed (1) / Tests 3 failed | 4 passed (7)', failure text 'AssertionError: expected {} to deeply equal { planId: 'rec-plan-1' }' (the empty request body IS the defect) and 'expected true to be false' on the non-default recordIdField refusal. GREEN after fix, verified at f08cfa8ec: new file + RecordDetailView.recordIdParamSeed.test.tsx + packages/core/src/actions/__tests__/recordIdParam.test.ts -> 'Test Files 3 passed (3) / Tests 23 passed (23)'. Full view suite: `pnpm exec vitest run packages/app-shell/src/views/` -> 'Test Files 272 passed (272) / Tests 2680 passed | 1 skipped (2681)'. No ablation run — red-before/green-after on the real defect is the stronger evidence here and was performed in that order. No existing test weakened, skipped or quarantined; the 6 pre-existing pins in the sibling file pass unchanged.",
  "gates": {
    "derivation": "Derived from the actual diff (2 app-shell src files + 1 changeset) against this repo's own workflows — lint.yml, ci.yml, changeset-*.yml, control-bytes.yml. scripts/pm/dispatch-gates.mjs is objectstack-scoped and does not exist here.",
    "run_locally_at_f08cfa8ec": {
      "turbo run type-check --filter @object-ui/app-shell": "PASS — 30 tasks successful (app-shell runs `tsc --noEmit && tsc -p tsconfig.test.json`, so the new test file compiles)",
      "turbo run lint --filter @object-ui/app-shell": "PASS — 0 errors, 2412 pre-existing no-explicit-any warnings (unchanged; no new as-any added)",
      "vitest packages/app-shell/src/views/": "PASS — 272 files / 2680 tests",
      "check-control-bytes": "PASS — scanned 4593 tracked text files",
      "check-lint-coverage": "PASS — 46/46 packages",
      "check-type-check-coverage": "PASS — 45/46 via type-check, 41/41 compile tests",
      "check-changeset-presence": "PASS — 2 released-package src files changed, 1 changeset declared",
      "check-changeset-no-major": "PASS — no major bump declared",
      "check-changeset-fixed": "PASS",
      "check:action-forward-parity": "PASS",
      "check:phantom-deps": "PASS",
      "check:self-import": "PASS",
      "check:spec-symbols": "PASS",
      "check:i18n-keys": "PASS",
      "check:i18n-drift": "PASS"
    },
    "left_to_ci": "The four Test (shard N/4) jobs run the full repo suite; I ran the app-shell view slice my diff implicates. Not implicated by this diff and not run: docs-links, skills-paths, published-dist, shadcn-check, performance-budget, live-e2e.",
    "manual_control_byte_sweep": "grep -naP '[\\x00-\\x08\\x0b\\x0c\\x0e-\\x1f\\x7f]' over all three changed files — clean"
  },
  "deviations": [
    "None on file surface: change is confined to RecordDetailView.tsx + a new RecordDetailView.*.test.tsx sibling. DeclaredActionsBar.tsx NOT touched — the second dispatch path the dispatch allowed for does not exist.",
    "Zone 3's suggested route (disable the affordance) was not taken; measurement contradicted its premise. Reasoning recorded above and in the PR body.",
    "No skip-changeset label applied or created (per dev contract, that label must never be minted in objectui); a real patch changeset is declared instead."
  ],
  "open_questions": [],
  "out_of_scope_findings": []
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/record-header-recordid-load-window-5176.md
+++ b/.changeset/record-header-recordid-load-window-5176.md
@@ -1,0 +1,22 @@
+---
+'@object-ui/app-shell': patch
+---
+
+Fix `record_header` actions dispatching without their record id while the record page is still loading.
+
+The record header does not wait for the page record: the skeleton's `isLoading`
+flag is cleared in a route-keyed microtask, but `pageRecord` only lands one
+`findOne` round-trip later, so the real header renders with every
+`record_header` action live for that whole window. An `api` action clicked
+inside it had no seed for its `recordIdParam` — the fallback chain stopped at
+`pageRecord` and never consulted the record id carried in the URL — so the
+request was dispatched naming no record and the backend rejected it with
+`missing_record_id`. Users saw a button that "did nothing until you clicked it
+again".
+
+The id the page is addressed by now seeds the parameter as a last resort, which
+is the same fallback this view's other dispatch paths already use. A stashed row
+and an explicit `recordId` override still take precedence, actions retargeting
+another object are still never handed this page's id, and an action keyed on a
+non-default `recordIdField` now takes a named refusal instead of emitting an
+under-specified request.

--- a/packages/app-shell/src/views/RecordDetailView.headerRecordIdLoadWindow.test.tsx
+++ b/packages/app-shell/src/views/RecordDetailView.headerRecordIdLoadWindow.test.tsx
@@ -1,0 +1,402 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * `record_header` actions must carry the record id even while the page record
+ * is still being fetched (objectui#5176).
+ *
+ * The header is NOT gated on the record load. `isLoading` — the only state the
+ * `<SkeletonDetail>` early return reads — is flipped false in a `queueMicrotask`
+ * keyed on the route, while `pageRecord` arrives one `dataSource.findOne`
+ * round-trip later. So for the whole duration of that fetch the page renders its
+ * real header, with every `record_header` action live and clickable, and
+ * `pageRecord` still `null`.
+ *
+ * In that window the `type:'api'` handler's `recordIdParam` seeding used to have
+ * no source at all. Its fallback chain was row -> literal `recordId` override ->
+ * `pageRecord`, and it stopped there — it never consulted `pureRecordId`, the id
+ * this page is *addressed by*, which is read straight off the URL and is
+ * therefore available on the very first render. Handed a `null` seed record,
+ * `resolveRecordIdParamSeed` deliberately abstains (`{}` — no value, and no
+ * error, because "no row context" is not that guard's business), so the caller
+ * injected nothing, raised nothing, and dispatched anyway. What went out was a
+ * mutation naming no record, and the backend answered `missing_record_id`.
+ *
+ * That is the user-visible "click does nothing, click again and it works": the
+ * second click lands after the fetch settled. The reporter's one-to-one
+ * correlation with a differently-shaped header DOM is real but is a co-symptom,
+ * not the cause — the same unresolved `pageRecord` that starves the seed also
+ * withholds the title/highlights chrome that makes the header render differently.
+ *
+ * The three sibling dispatch paths in this file (the legacy `dataSource.update`
+ * default branch, the flow trigger, and the server-action bridge) all already
+ * fall back to `pureRecordId`. This pins the api path to the same source, and
+ * pins the one case the URL id cannot serve — a non-default `recordIdField` —
+ * to a refusal rather than an under-specified request.
+ */
+
+import * as React from 'react';
+import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
+import { render, waitFor, act, cleanup } from '@testing-library/react';
+import { MemoryRouter } from 'react-router-dom';
+
+const authFetchSpy = vi.fn(async () =>
+  new Response(JSON.stringify({ data: [] }), {
+    status: 200,
+    headers: { 'content-type': 'application/json' },
+  }),
+);
+vi.mock('@object-ui/auth', () => ({
+  useAuth: () => ({ user: { id: 'u1', name: 'Ada', image: null }, activeOrganization: null }),
+  createAuthenticatedFetch: () => authFetchSpy,
+}));
+
+vi.mock('@object-ui/collaboration', () => ({
+  useRecordPresence: () => [],
+  PresenceAvatars: () => null,
+}));
+
+vi.mock('sonner', () => ({
+  toast: Object.assign(vi.fn(), {
+    success: vi.fn(),
+    error: vi.fn(),
+    info: vi.fn(),
+    warning: vi.fn(),
+    loading: vi.fn(),
+    dismiss: vi.fn(),
+  }),
+}));
+
+vi.mock('./ActionConfirmDialog', () => ({ ActionConfirmDialog: () => null }));
+vi.mock('./ActionParamDialog', () => ({ ActionParamDialog: () => null }));
+vi.mock('./ActionResultDialog', () => ({ ActionResultDialog: () => null }));
+vi.mock('./FlowRunner', () => ({ FlowRunner: () => null }));
+vi.mock('./MetadataInspector', () => ({
+  MetadataPanel: () => null,
+  useMetadataInspector: () => ({ showDebug: false, toggle: () => {} }),
+}));
+
+vi.mock('../hooks/useActionModal', () => ({
+  useActionModal: () => ({
+    modalHandler: vi.fn(async () => ({ success: true })),
+    modalElement: null,
+    closeModal: () => {},
+    resolveModalTarget: vi.fn(async () => null),
+  }),
+}));
+
+vi.mock('../utils/consoleServerAction', () => ({
+  createConsoleServerActionHandler: () => vi.fn(async () => ({ success: true })),
+}));
+
+// Same capture posture as the sibling `recordIdParamSeed` pin — keep the real
+// provider, record every handler set it is handed. The discriminator differs
+// and that difference is the whole point of this file: the sibling waits for
+// the capture whose context carries the LOADED record, which by construction
+// sits after the window under test here.
+const captured: Array<{ handlers: Record<string, (action: any) => Promise<any>>; context: any }> = [];
+vi.mock('@object-ui/react', async (importOriginal) => {
+  const actual = await importOriginal<typeof import('@object-ui/react')>();
+  return {
+    ...actual,
+    ActionProvider: (props: any) => {
+      captured.push({ handlers: props.handlers, context: props.context });
+      return React.createElement(actual.ActionProvider as any, props);
+    },
+    SchemaRenderer: () => null,
+  };
+});
+
+import { MetadataCtx } from '@object-ui/react';
+import { RecordDetailView } from './RecordDetailView';
+
+const OBJECT_NAME = 'os_production_plan';
+const RECORD_ID = 'rec-plan-1';
+
+const OBJECTS = [
+  {
+    name: OBJECT_NAME,
+    label: 'Production Plan',
+    fields: {
+      id: { type: 'text', label: 'Id' },
+      name: { type: 'text', label: 'Name' },
+      code: { type: 'text', label: 'Code' },
+    },
+  },
+];
+
+const METADATA = {
+  objects: OBJECTS,
+  pages: [],
+  loading: false,
+  error: null,
+  refresh: async () => {},
+  invalidate: () => {},
+  ensureType: async () => [],
+  getItem: async () => null,
+  getItemsByType: () => [],
+} as any;
+
+/**
+ * A data source whose `findOne` is held open until the test releases it. This
+ * is what makes the 1-in-10 timing report a deterministic case: the failure
+ * window is not sampled, it is held.
+ */
+function makeDeferredDataSource() {
+  let release!: (rec: unknown) => void;
+  const settled = new Promise<unknown>((resolve) => {
+    release = resolve;
+  });
+  const ds = {
+    find: vi.fn(async () => ({ data: [] })),
+    findOne: vi.fn(() => settled),
+    create: vi.fn(async () => ({})),
+    update: vi.fn(async () => ({})),
+    delete: vi.fn(async () => ({})),
+  } as any;
+  return { ds, release: () => release({ id: RECORD_ID, name: 'Plan A' }) };
+}
+
+/**
+ * The record page's own provider, captured WHILE the record load is still in
+ * flight: `context.objectName` identifies this page, and the absence of a
+ * record id in the context is precisely the half-loaded state
+ * (`record: pageRecord || {}`).
+ */
+function loadWindowCapture() {
+  return [...captured]
+    .reverse()
+    .find((c) => c.handlers && c.context?.objectName === OBJECT_NAME && !c.context?.record?.id);
+}
+
+async function mountMidLoadAndCaptureHandlers(ds: any) {
+  render(
+    <MemoryRouter initialEntries={[`/app/demo/${OBJECT_NAME}/${RECORD_ID}`]}>
+      <MetadataCtx.Provider value={METADATA}>
+        <RecordDetailView
+          dataSource={ds}
+          objects={OBJECTS}
+          onEdit={() => {}}
+          objectNameOverride={OBJECT_NAME}
+          recordIdOverride={RECORD_ID}
+          embedded
+        />
+      </MetadataCtx.Provider>
+    </MemoryRouter>,
+  );
+  await waitFor(() => expect(loadWindowCapture()).toBeTruthy());
+  authFetchSpy.mockClear();
+  return loadWindowCapture()!;
+}
+
+beforeEach(() => {
+  cleanup();
+  captured.length = 0;
+  authFetchSpy.mockClear();
+  vi.stubGlobal(
+    'fetch',
+    vi.fn(async () =>
+      new Response(JSON.stringify({ data: [] }), {
+        status: 200,
+        headers: { 'content-type': 'application/json' },
+      }),
+    ),
+  );
+});
+
+afterEach(() => {
+  vi.unstubAllGlobals();
+});
+
+describe('RecordDetailView record_header actions during the page-record load window (objectui#5176)', () => {
+  it('the header renders live actions while the record is still loading', async () => {
+    // The premise the whole card rests on. If the header were gated on the
+    // record load there would be no window to dispatch in, and the reported
+    // failure would be unreachable.
+    const { ds } = makeDeferredDataSource();
+    const capture = await mountMidLoadAndCaptureHandlers(ds);
+
+    expect(ds.findOne).toHaveBeenCalled();
+    expect(typeof capture.handlers.api).toBe('function');
+    // Half-loaded by construction: the provider is up, the record is not.
+    expect(capture.context?.record?.id).toBeUndefined();
+  });
+
+  it('seeds recordIdParam from the URL record id when the page record has not resolved', async () => {
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    let r: any;
+    await act(async () => {
+      // A record_header action exactly as authored: no `_rowRecord` stash, no
+      // literal `recordId` override, `recordIdField` defaulting to `id`.
+      r = await handlers.api({
+        name: 'archive_plan',
+        type: 'api',
+        method: 'POST',
+        target: `/api/v1/data/${OBJECT_NAME}/archive`,
+        recordIdParam: 'planId',
+      });
+    });
+
+    expect(r.success).toBe(true);
+    expect(authFetchSpy).toHaveBeenCalledTimes(1);
+    const [, init] = authFetchSpy.mock.calls[0] as unknown as [string, RequestInit];
+    const body = JSON.parse(String(init.body));
+    // The defect: this key was absent, and the request went out anyway.
+    expect(body).toMatchObject({ planId: RECORD_ID });
+
+    release();
+  });
+
+  it('injects identically before and after the record resolves', async () => {
+    // The triage floor's first limb, stated as one assertion: the two header
+    // render states are not allowed to disagree about the request body.
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers: midLoadHandlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    const action = {
+      name: 'archive_plan',
+      type: 'api',
+      method: 'POST',
+      target: `/api/v1/data/${OBJECT_NAME}/archive`,
+      recordIdParam: 'planId',
+    };
+
+    await act(async () => {
+      await midLoadHandlers.api({ ...action });
+    });
+    const [, midInit] = authFetchSpy.mock.calls[0] as unknown as [string, RequestInit];
+    const midBody = JSON.parse(String(midInit.body));
+
+    authFetchSpy.mockClear();
+    await act(async () => {
+      release();
+    });
+    await waitFor(() =>
+      expect(
+        [...captured].reverse().find((c) => c.handlers && c.context?.record?.id === RECORD_ID),
+      ).toBeTruthy(),
+    );
+    const loadedHandlers = [...captured]
+      .reverse()
+      .find((c) => c.handlers && c.context?.record?.id === RECORD_ID)!.handlers;
+
+    await act(async () => {
+      await loadedHandlers.api({ ...action });
+    });
+    const [, loadedInit] = authFetchSpy.mock.calls[0] as unknown as [string, RequestInit];
+    const loadedBody = JSON.parse(String(loadedInit.body));
+
+    expect(midBody).toEqual(loadedBody);
+    expect(midBody).toMatchObject({ planId: RECORD_ID });
+  });
+
+  it('refuses rather than under-specifying when recordIdField names a field the URL id cannot supply', async () => {
+    // The floor's second limb, for the residue the first cannot cover: the URL
+    // carries the record's *id*, so an action keyed on some other field still
+    // has no seed mid-load. It must refuse — never emit the body-incomplete
+    // request this card is about.
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    let r: any;
+    await act(async () => {
+      r = await handlers.api({
+        name: 'link_code',
+        label: 'Link by Code',
+        type: 'api',
+        method: 'POST',
+        target: `/api/v1/data/${OBJECT_NAME}/link`,
+        recordIdParam: 'targetCode',
+        recordIdField: 'code',
+      });
+    });
+
+    expect(r.success).toBe(false);
+    expect(r.error).toContain('Link by Code');
+    expect(r.error).toContain('code');
+    expect(authFetchSpy).not.toHaveBeenCalled();
+
+    release();
+  });
+
+  it('still prefers a stashed row over the URL id', async () => {
+    // The new fallback is a LAST resort: it must not outrank the row a
+    // related-list dispatch stashes, which may name a different record.
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    let r: any;
+    await act(async () => {
+      r = await handlers.api({
+        name: 'archive_plan',
+        type: 'api',
+        method: 'POST',
+        target: `/api/v1/data/${OBJECT_NAME}/archive`,
+        recordIdParam: 'planId',
+        params: { _rowRecord: { id: 'row-child-7' } },
+      });
+    });
+
+    expect(r.success).toBe(true);
+    const [, init] = authFetchSpy.mock.calls[0] as unknown as [string, RequestInit];
+    expect(JSON.parse(String(init.body))).toMatchObject({ planId: 'row-child-7' });
+
+    release();
+  });
+
+  it('still prefers a literal recordId override over the URL id', async () => {
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    let r: any;
+    await act(async () => {
+      r = await handlers.api({
+        name: 'archive_plan',
+        type: 'api',
+        method: 'POST',
+        target: `/api/v1/data/${OBJECT_NAME}/archive`,
+        recordIdParam: 'planId',
+        recordId: 'explicit-plan-9',
+      } as any);
+    });
+
+    expect(r.success).toBe(true);
+    const [, init] = authFetchSpy.mock.calls[0] as unknown as [string, RequestInit];
+    expect(JSON.parse(String(init.body))).toMatchObject({ planId: 'explicit-plan-9' });
+
+    release();
+  });
+
+  it('does not seed this page id into an action retargeting another object', async () => {
+    // Same guard `interpolationRecord` uses: a child-object action must never
+    // be handed the parent page's id. Mid-load it has no seed, so it refuses.
+    const { ds, release } = makeDeferredDataSource();
+    const { handlers } = await mountMidLoadAndCaptureHandlers(ds);
+
+    let r: any;
+    await act(async () => {
+      r = await handlers.api({
+        name: 'archive_other',
+        label: 'Archive Other',
+        type: 'api',
+        method: 'POST',
+        objectName: 'other_object',
+        target: '/api/v1/data/other_object/archive',
+        recordIdParam: 'planId',
+        params: { _rowRecord: { name: 'no id here' } },
+      });
+    });
+
+    expect(r.success).toBe(false);
+    expect(authFetchSpy).not.toHaveBeenCalled();
+
+    release();
+  });
+});

--- a/packages/app-shell/src/views/RecordDetailView.tsx
+++ b/packages/app-shell/src/views/RecordDetailView.tsx
@@ -638,16 +638,42 @@ export function RecordDetailView({ dataSource, objects, onEdit, objectNameOverri
           // refusal only once every source has been tried, instead of the
           // silent drop this call site used to fall back to
           // (objectstack#8018, objectui#4669).
+          //
+          // objectui#5176 — the page-record fallback is not enough on its own,
+          // because the header does not wait for the page record. `isLoading`
+          // (the only state the <SkeletonDetail> early return reads) is flipped
+          // false in a route-keyed microtask, while `pageRecord` lands one
+          // `findOne` round-trip later; for that whole window the real header
+          // renders with every `record_header` action live. Clicking inside it
+          // seeded from `null`, which `resolveRecordIdParamSeed` deliberately
+          // abstains on (no value AND no error — "no row context" is not that
+          // guard's business), so the request went out naming no record and the
+          // backend answered `missing_record_id`. That is the reported "click
+          // does nothing, click again and it works".
+          //
+          // `pureRecordId` is the id this page is ADDRESSED by — read off the
+          // URL, present on the first render, and already the fallback the
+          // other three dispatch paths in this file use. Seeding a minimal row
+          // from it makes this path agree with them. It stays a LAST resort
+          // (row and override still outrank it) and keeps the same
+          // retarget guard `interpolationRecord` uses, so a child-object action
+          // is never handed the parent page's id. An action keyed on some other
+          // field via `recordIdField` gets a row without that key, which is the
+          // helper's named refusal — deliberately, since the URL carries the
+          // record's id and nothing else: refusing beats under-specifying.
           const recordIdOverride = (action as { recordId?: unknown }).recordId;
           const rowHasValue = !!rowRecord && rowRecord[rowField] != null;
           if (!rowHasValue && recordIdOverride != null) {
             body[action.recordIdParam] = recordIdOverride;
           } else {
+            const targetsThisRecord = !action.objectName || action.objectName === objectName;
+            const thisRecordSeed: Record<string, any> | undefined = targetsThisRecord
+              ? ((pageRecord as Record<string, any> | undefined)
+                  ?? (pureRecordId != null ? { id: pureRecordId } : undefined))
+              : undefined;
             const seedRecord: Record<string, any> | undefined = rowHasValue
               ? rowRecord
-              : ((!action.objectName || action.objectName === objectName
-                  ? (pageRecord as Record<string, any> | undefined)
-                  : undefined) ?? rowRecord);
+              : (thisRecordSeed ?? rowRecord);
             const seed = resolveRecordIdParamSeed(action, seedRecord);
             if (seed.error) return { success: false, error: seed.error };
             if (seed.value !== undefined) body[action.recordIdParam] = seed.value;

